plugin/references.py

```diff
@@ -57,44 +58,85 @@ def run(self, edit, event=None):
-                    'force': True,
-                    'scroll_to_end': True
-                })
-            panel.set_read_only(True)
-
-        else:
+        references_count = len(response)
+        # return if there are no references
+        if references_count < 1:
             window.run_command("hide_panel", {"panel": "output.references"})
             window.status_message("No references found")
+            return
+
+        word = self.view.substr(self.view.word(pos))
+
+        base_dir = get_project_path(window)
+        formated_references = self._get_formated_references(response, base_dir)
+
+        panel = ensure_references_panel(window)
+        panel.settings().set("result_base_dir", base_dir)
+
+        panel.set_read_only(False)
+        panel.run_command("lsp_clear_panel")
+        window.run_command("show_panel", {"panel": "output.references"})
+        panel.run_command('append', {
+            'characters': "{} references for '{}'\n\n{}".format(references_count, word, formated_references),
+            'force': True,
+            'scroll_to_end': False
+        })
+
+        # highlight all word occurrences
+        regions = panel.find_all(r"\b{}\b".format(word))
+        panel.add_regions('ReferenceHighlight', regions, 'comment', flags=sublime.DRAW_OUTLINED)
+        panel.set_read_only(True)
 
     def want_event(self):
         return True
 
-
-def format_reference(reference, base_dir):
-    start = Point.from_lsp(reference.get('range').get('start'))
-    file_path = uri_to_filename(reference.get("uri"))
-    relative_file_path = os.path.relpath(file_path, base_dir)
-    return " ◌ {} {}:{}".format(relative_file_path, start.row + 1, start.col + 1)
+    def _get_formated_references(self, references: 'List[Dict]', base_dir) -> str:
+        grouped_references = self._group_references_by_file(references, base_dir)
+        formated_references = self._format_references(grouped_references)
+        return formated_references
+
+    def _group_references_by_file(self, references, base_dir):
+        """ Return a dictionary that groups references by the file it belongs.
+        Example:
+        {
+            'main.py': [
+                { Point, 'from .src.one import one' },
+                { Point, 'two = 1 + one()' }
+            ],
+            'src/one.py': [
+               { Point, 'def one():' }
+            ]
+        }"""
+        dict = {}  # type: Dict[str, List[Dict]]
+        for reference in references:
+            file_path = uri_to_filename(reference.get("uri"))
+            relative_file_path = os.path.relpath(file_path, base_dir)
+
+            point = Point.from_lsp(reference.get('range').get('start'))
+            # get line of the reference, to showcase its use
+            reference_line = linecache.getline(file_path, point.row + 1).strip()
+
+            # we don't want to cache the line, we always want to get fresh data
+            linecache.clearcache()
```

Comment:
Should `clearcache` be called in a loop? 
Is clearing the line cache necessary for "grouping references" or should it be done in another part of the code ?
I assume the point is to use _some_ caching per references request?

Reply:
`clearcache` should't be inside of a loop.
I need to call it only once.